src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueServer.scala

```diff
@@ -0,0 +1,144 @@
+
+import implicits._
+
+/**
+ * [[https://github.com/grpc/grpc-java/blob/v0.15.0/examples/src/main/java/io/grpc/examples/helloworld/HelloWorldServer.java]]
+ */
+object OpaqueServer {
+  private val logger = Logger.getLogger(classOf[OpaqueServer].getName)
+
+  def main(args: Array[String]): Unit = {
+    val server = new OpaqueServer(ExecutionContext.global)
+    server.start()
+    server.blockUntilShutdown()
+    println("Print hello world server")
+  }
+
+  private val port = 50051
+}
+
+class OpaqueServer(executionContext: ExecutionContext) { self =>
+  private[this] var server: Server = null
+  private[this] var spark: SparkSession = null
+
+  private def start(): Unit = {
+
+    // initialize SparkSession and ensure remote attestation works
+    spark = SparkSession.builder()
+      .appName("QEDBenchmark")
+      .getOrCreate()
+    Utils.initSQLContext(spark.sqlContext)
+
+    // Create spark.sql table for testing purposes
+    val data = for (i <- 0 until 256) yield (i, i*2, 1)
+    val df = spark.createDataFrame(data).toDF("first", "second", "third")
+    val dfEncrypted = df.encrypted
+    dfEncrypted.createTempView("Test")
+
+    // Create rpc Server and make it listen
+    server = ServerBuilder.forPort(OpaqueServer.port).addService(new GreeterImpl()).build.start
+    OpaqueServer.logger.info("Server started, listening on " + OpaqueServer.port)
+    sys.addShutdownHook {
+      System.err.println("*** shutting down gRPC server since JVM is shutting down")
+      self.stop()
+      System.err.println("*** server shut down")
+      
+      // Drop test table
+      spark.catalog.dropTempView("Test")
+    }
+  }
+
+  private def stop(): Unit = {
+    if (server != null) {
+      server.shutdown()
+    }
+  }
+
+  private def blockUntilShutdown(): Unit = {
+    if (server != null) {
+      server.awaitTermination()
+    }
+  }
+
+  private class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+    override def sayHello(req: Ra.HelloRequest, responseObserver: io.grpc.stub.StreamObserver[Ra.HelloReply]) = {
+      val reply = Ra.HelloReply.newBuilder().setMessage("Hello " + req.getName()).build();
+      responseObserver.onNext(reply);
+      responseObserver.onCompleted();
+    }
+
+    override def getRA(req: Ra.RARequest, responseObserver: io.grpc.stub.StreamObserver[Ra.RAReply]) = {
+      
+      val (enclave, eid) = Utils.initEnclave()
+      val msg1 = enclave.GenerateReport(eid)
+
+      val byteString: ByteString = ByteString.copyFrom(msg1)
+
+      val reply = Ra.RAReply.newBuilder()
+	.setReport(byteString)
+	.setSuccess(true)
+	.build();
+
+      responseObserver.onNext(reply);
+      responseObserver.onCompleted();
+    }
+
+    override def sendQuery(req: Ra.QueryRequest, responseObserver: io.grpc.stub.StreamObserver[Ra.QueryReply]) = {
+      var reply = Ra.QueryReply.newBuilder().build();
+      try {
+	// Prepare items for setting data reply
+        val result = spark.sql(req.getSqlQuery())
```

Comment:
Have you also tried testing this for loading saved tables, in particular in EncryptedSource format? i.e. to simulate the context, you can create an encrypted dataframe in Opaque and save it to disk (all at the server), then try to load that dataframe and manipulate it over gRPC.